ARROW-5525: [C++] Add Continuous Fuzzing Integration setup with Fuzzit

yevgenypats opened the pull request:
Adds fuzzit test to docker-compose and dev/tasks/tests.yml

crepererum commented:
FYI: I'll work together w/ @yevgenypats to address the following issues:

- **fuzzer binaries should be statically linked:** this will simplify the work for Fuzzit and other continuous fuzzing services
- **accesss clarification:** since the fuzzing results (crashes) may contain security-sensitive information, we should think if these are publicly available or if only arrow maintainers should see them
- **corpus:** fuzzing w/o a proper seed corpus is kinda wasteful. we should have a corpus that containers
  - previous found and fixed crashes
  - some valid arrow messages (simple and complex ones), I think the integration test files will work here

crepererum commented:
## Static Linking
The binary is statically linked against the arrow code, but dynamically linked against the system libs. We cannot completely remove this restrictions [due to address sanitizer](https://clang.llvm.org/docs/AddressSanitizer.html#limitations), but we can try to limit the host-library interaction:

```bash
cmake \
    -DARROW_FUZZING=ON \
    -DARROW_USE_ASAN=ON \
    -DCMAKE_BUILD_TYPE=RelWithDebInfo \
    -DARROW_DEPENDENCY_SOURCE=BUNDLED \
    ..
```

With this, I have the following libs left:

```text
/lib64/ld-linux-x86-64.so.2
libboost_filesystem.so.1.69.0
libboost_regex.so.1.69.0
libboost_system.so.1.69.0
libc.so.6
libcrypto.so.1.1
libdl.so.2
libgcc_s.so.1
libicudata.so.64
libicui18n.so.64
libicuuc.so.64
libm.so.6
libpthread.so.0
librt.so.1
libstdc++.so.6
linux-vdso.so.1
```

Sadly, I wasn't able to build arrow with a vendored boost. And I think the sanitizer library will also be loaded at runtime.

## Corpus
Is now hosted under https://github.com/apache/arrow-testing/tree/master/data/arrow-ipc. I'll try to extend this in the coming week(s).

## Known Bugs
I currently have some bugs that I have discovered already by fuzzing locally and that I'm trying to fix them. This shouldn't prevent anyone from running a fuzzer, but be warned that you may run into them rather quickly.

yevgenypats commented:
@crepererum Thank you for the detailed reply! I think this is good enough. I'll update in the next couple of days with the first version of the integration.

bkietz commented on .gitmodules:
```diff
@@ -1,6 +1,3 @@
 [submodule "cpp/submodules/parquet-testing"]
 	path = cpp/submodules/parquet-testing
 	url = https://github.com/apache/parquet-testing.git
-[submodule "testing"]
```
Why delete this submodule?

yevgenypats commented:
Hi, A mistake in rebase. It's still wip. I'll fix this.

On Mon, Jun 17, 2019, 9:59 PM Benjamin Kietzman <notifications@github.com>
wrote:

> *@bkietz* commented on this pull request.
> ------------------------------
>
> In .gitmodules
> <https://github.com/apache/arrow/pull/4504#discussion_r294470252>:
>
> > @@ -1,6 +1,3 @@
>  [submodule "cpp/submodules/parquet-testing"]
>  	path = cpp/submodules/parquet-testing
>  	url = https://github.com/apache/parquet-testing.git
> -[submodule <https://github.com/apache/parquet-testing.git-%5Bsubmodule> "testing"]
>
> Why delete this submodule?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/arrow/pull/4504?email_source=notifications&email_token=AD52CDVGW5MVSTGRPFDKADDP27NJHA5CNFSM4HWJSMS2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3YTAFY#pullrequestreview-250687511>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD52CDUZL6OC3ZKQD6VSV73P27NJHANCNFSM4HWJSMSQ>
> .
>

yevgenypats commented:
@crepererum The pull-request is almost ready but I need a bit of help on your end.
 
As you said it's already found a crash. I added you to the [apache/arrow](https://app.fuzzit.dev/admin/yMxZh42xl9qy6bvg3EiJ/dashboard) organization so you can take a look.

Currently the Fuzzit integration is configured to run on every PR and commit which is too much and might just spam your inbox because of PRs that are not ready yet.

My suggestion is to create a cron job in travis and run this step on a daily basis from master - kind like done in [systemd](https://github.com/systemd/systemd/blob/master/.travis.yml).

Let me know what you think.

crepererum commented:
> My suggestion is to create a cron job in travis and run this step on a daily basis from master - kind like done in systemd.

Sounds sane. @wesm would it be OK to add a cron job to travis?

yevgenypats commented:
No problem at all. Can you show me an example of a job running there? Also
of course let's make sure it possible to run nightly.

On Wed, Jun 19, 2019, 9:58 AM Wes McKinney <notifications@github.com> wrote:

> Oof I didn't see that you were setting this up in Travis CI. Can you
> configure it as a docker-compose job instead? We're already spilling over
> what Travis can do for us -- this seems like something we'd rather run on
> the Buildbot physical infra (if it's Dockerized, it can be run anywhere
> where there is Docker), but I'm not sure how to set up a nightly job there
> yet cc @kszucs <https://github.com/kszucs> @nealrichardson
> <https://github.com/nealrichardson>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/arrow/pull/4504?email_source=notifications&email_token=AD52CDWMNW65XIJE5AMUWB3P3I3QFA5CNFSM4HWJSMS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYB6RDY#issuecomment-503572623>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD52CDSQYVEO3SG3CARL5YDP3I3QFANCNFSM4HWJSMSQ>
> .
>

wesm commented:
Take a look at jobs in 

https://github.com/apache/arrow/blob/master/docker-compose.yml

An example is `iwyu` which uses the clang toolchain to build and run include-what-you-use

To add a nightly job, add an entry to https://github.com/apache/arrow/blob/master/dev/tasks/tests.yml#L21

Here is an example of a Travis CI configuration that can be used to set up a nightly cron job

https://github.com/apache/arrow/blob/master/dev/tasks/nightlies.sample.yml

Here's an example of a nightly run https://github.com/ursa-labs/crossbow/branches/all?page=1&query=nightly-481&utf8=%E2%9C%93

@kszucs is working to better document some of these things cc @nealrichardson

yevgenypats commented:
@crepererum can you give me a hand here? Essentially my script can run in any docker which has a clang>6.0. But i'm not quite sure I understood the steps @wesm provided. Are you familiar with that? if yes, it will save me some time.

crepererum commented:
No, that build system setup is new to me (but I never actually modified the arrow build infrastructure before). I would say:

1. Create a `fuzzing` folder in https://github.com/apache/arrow/tree/master/dev/tasks
2. Take the [`nightlies.sample.yml`](https://github.com/apache/arrow/blob/master/dev/tasks/nightlies.sample.yml) template and merge it with your `.travis.yml` modifications and store it in the mentioned `fuzzing` folder.
3. Add the job [here](https://github.com/apache/arrow/blob/master/dev/tasks/tests.yml#L21) and [here](https://github.com/apache/arrow/blob/master/dev/tasks/tasks.yml#L57).

yevgenypats commented:
@wesm Where can I see my job running? so I can test that the new integration really works?

wesm commented:
You can see nightly builds in e.g. https://github.com/ursa-labs/crossbow/branches/all?utf8=%E2%9C%93&query=nightly-488

built by Travis CI cron job

https://github.com/ursa-labs/crossbow/blob/trigger-nightly-builds/.travis.yml

yevgenypats commented:
@wesm you mentioned the buildbot physical infra. how is it connected totravis CI cron jobs?

wesm commented:
They aren't related, but we can use ursabot to trigger these Dockerized Crossbow jobs. I'll show an example invocation

wesm commented:
@ursabot crossbow test docker-continuous-fuzzing

wesm commented:
@ursabot crossbow test docker-cpp

yevgenypats commented:
@wesm, sorry for nudging you but can you give me the exact instruction on where to add my code.

wesm commented:
Take a look at the `docker-compose.yml` in the root of the project. You have to add an entry there

yevgenypats commented:
@wesm I've added an entry to docker-compose and tested that this works locally. I've also added an entry to tests.yaml. Now how I make sure it is running on a nightly basis?

crepererum commented:
@yevgenypats BTW: I'm offline for a week and will be available at the 8th of July again.

wesm commented:
Thanks for your patience

yevgenypats commented:
@wesm sure. thx!

wesm commented:
@yevgenypats can you create an account on http://issues.apache.org/jira so I can assign this issue to you?

yevgenypats commented:
@wesm Yes, my username is yevgenyp. Also, why we didn't see the nightly yet? Following up on my previous comment - how we make sure it will run on a nightly basis? maybe I missed something.

wesm commented:
The nightly won't run until the patch is in master

yevgenypats commented:
Yeah, got it one second after posting the question, thx.

yevgenypats commented:
@wesm Looks like the job was triggered successfully! Though there is one fix needs to be done. looks like the CIRCLE_SHA1 variable and CIRCLE_BRANCH that I'm using are empty. how do I access the branch and commit info from the `docker_build_and_fuzzit.sh` script?

wesm commented:
@kszucs do you know the answer to this?

yevgenypats commented:
@wesm can you also retweet this for our OSS contrib, this will help a lot.
Thx! https://twitter.com/fuzzitdev/status/1148307353990828032?s=09



On Wed, Jul 10, 2019, 11:04 PM Wes McKinney <notifications@github.com>
wrote:

> @kszucs <https://github.com/kszucs> do you know the answer to this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/arrow/pull/4504?email_source=notifications&email_token=AD52CDW6WS4LE5OPZTNLBWDP6Y6FHA5CNFSM4HWJSMS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZUSO3A#issuecomment-510207852>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD52CDVRFSOSSFSVALZ22YTP6Y6FHANCNFSM4HWJSMSQ>
> .
>

kszucs commented:
@yevgenypats you'll need to proxy it through docker-compose, add something like the following to the [fuzzit definition](https://github.com/apache/arrow/blob/master/docker-compose.yml#L430):

```yml
environment:
  CI_ARROW_SHA: ${CI_ARROW_SHA}
  CI_ARROW_BRANCH: ${CI_ARROW_BRANCH}
```
and export `CI_ARROW_SHA`, `CI_ARROW_BRANCH` int the crossbow [docker-test templates](https://github.com/apache/arrow/tree/master/dev/tasks/docker-tests).

In case of circleci export as `CIRCLE_SHA1` and `CIRCLE_BRANCH` respectively.

yevgenypats commented:
@kszucs can you give me an example of how the export should look like in [circle.linux.yml](https://github.com/apache/arrow/blob/master/dev/tasks/docker-tests/circle.linux.yml)

yevgenypats commented:
@wesm @crepererum @kszucs  is this still relevant? if yes, we need to fix the branch name + I'll really appreciate a [retweet](https://twitter.com/fuzzitdev/status/1148307353990828032) for our contribution of free compute power. Cheers.

wesm commented:
Is there a JIRA about fixing it?

crepererum commented:
I've created [ARROW-5959](https://issues.apache.org/jira/browse/ARROW-5959) and assigned it to @yevgenypats for the time being.
